staging/src/k8s.io/code-generator/pkg/util/comments.go

```
@@ -0,0 +1,88 @@
+*/
+
+package util
+
+import (
+	"k8s.io/gengo/v2"
+)
+
+// ExtractCommentTagsWithoutArguments parses comments for special metadata tags. The
+// marker argument should be unique enough to identify the tags needed, and
+// should not be a marker for tags you don't want, or else the caller takes
+// responsibility for making that distinction.
+//
+// The tagNames argument is a list of specific tags being extracted. If this is
+// nil or empty, all lines which match the marker are considered.  If this is
+// specified, only lines with begin with marker + one of the tags will be
+// considered.  This is useful when a common marker is used which may match
+// lines which fail this syntax (e.g. which predate this definition).
+//
+// This function looks for input lines of the following forms:
+//   - 'marker' + "key=value"
+//   - 'marker' + "key()=value"
+//   - 'marker' + "key(arg)=value"
+//
+// The arg is optional.  This function only consider tags with no arguments specified
+// (either as "key=value" or as // "key()=value").
+//
+// The value is optional.  If not specified, the resulting Tag will have "" as
+// the value.
+//
+// Tag comment-lines may have a trailing end-of-line comment.
+//
+// The map returned here is keyed by the Tag's name without args.
+//
+// A tag can be specified more than one time and all values are returned.  If
+// the resulting map has an entry for a key, the value (a slice) is guaranteed
+// to have at least 1 element.
+//
+// Example: if you pass "+" for 'marker', and the following lines are in
+// the comments:
+//
+//	+foo=val1   // foo
+//	+bar
+//	+foo=val2   // also foo
+//	+foo()=val3 // still foo
+//	+baz="qux"
+//	+foo(arg)   // tags with arguments are ignored
+//
+// Then this function will return:
+//
+//	map[string][]string{"foo":{"val1", "val2", "val3"}, "bar": {""}, "baz": {`"qux"`}}
+func ExtractCommentTagsWithoutArguments(marker string, tagNames []string, lines []string) (map[string][]string, error) {
```

Could you help me to understand why this is needed for the above generator?
If it is common enough, I suggest to put it inside gengo/v2 than import here to use.

---

The existing generator tags are not function-style, so the behavior of function-style usages is undefined. I propose we simply ignore tags that include arguments, but I’m happy to adjust the implementation if we decide to handle them differently.

This helper function helps reduce boilerplate by filtering and collecting parsed tags. It’s probably too specific for gengo/v2, but I’d be glad to submit it there if you think it might be useful.

| Usage | Proposed Effect |
| ------- | ----------------- |
| `+k8s:deepcopy-gen:nonpointer-interfaces=true` | Accepted |
| `+k8s:deepcopy-gen:nonpointer-interfaces()=true` | Accepted |
| `+k8s:deepcopy-gen:nonpointer-interfaces(a=1)=true` | Ignored |

What are your thoughts?

cc @jpbetz

---

Let's be strict and have the code generator fail with an error if an unexpected argument (`(a=1)` case) is present.

---

Sounds good. I can tweak that.

---

/lgtm